### PR TITLE
feat: make Nats jetstream Source work in parallel

### PR DESCRIPTION
**Pull request by tabVersion**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

tracked by https://github.com/risingwavelabs/risingwave/issues/18876

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by tabVersion** [comment]
test against main (b4bca5724e7acb9c39919aaa188ed699a10eb1f4)

load from Nats Jetstream 1 million message. 

---

* main + ack_policy: all

`create table t (a int ) with ( connector = 'nats', server_url='127.0.0.1:4222', subject='testsubject', connect_mode='plain', consumer.durable_name = 'main_all_1', consumer.ack_policy = 'all', stream='teststream', consumer.max_ack_pending = '100000') format plain encode json;`

image


* main + ack_policy: explicit

`create table t (a int ) with ( connector = 'nats', server_url='127.0.0.1:4222', subject='testsubject', connect_mode='plain', consumer.durable_name = 'main_explicit_1', consumer.ack_policy = 'explicit', stream='teststream', consumer.max_ack_pending = '100000') format plain encode json;`

image

* main + ack_policy: none

`create table t (a int ) with ( connector = 'nats', server_url='127.0.0.1:4222', subject='testsubject', connect_mode='plain', consumer.durable_name = 'main_none_4', consumer.ack_policy = 'none', stream='teststream', consumer.max_ack_pending = '100000') format plain encode json;`

image

---

* 4 parallel + ack_policy: all

`set streaming_parallelism to 4; create table t (a int ) with ( connector = 'nats', server_url='127.0.0.1:4222', subject='testsubject', connect_mode='plain', consumer.durable_name = 'branch_all_3', consumer.ack_policy = 'all', stream='teststream', consumer.max_ack_pending = '100000') format plain encode json;`

image

* 4 parallel + ack_policy: explicit

`set streaming_parallelism to 4; create table t (a int ) with ( connector = 'nats', server_url='127.0.0.1:4222', subject='testsubject', connect_mode='plain', consumer.durable_name = 'branch_explicit_3', consumer.ack_policy = 'explicit', stream='teststream', consumer.max_ack_pending = '100000') format plain encode json;`

image


* 4 parallel + ack_policy: none

`set streaming_parallelism to 4; create table t (a int ) with ( connector = 'nats', server_url='127.0.0.1:4222', subject='testsubject', connect_mode='plain', consumer.durable_name = 'branch_none_4', consumer.ack_policy = 'none', stream='teststream', consumer.max_ack_pending = '100000') format plain encode json;`

image

**Comment by tabVersion** [comment]
perf test using EC2

setup:

* Nats JS on EC2_1 (5,000,000 messages, file storage, GP3)
*  RisingWave on EC2_2

---

test against main (https://github.com/risingwavelabs/risingwave/commit/b4bca5724e7acb9c39919aaa188ed699a10eb1f4)

* ack all

```
create table t ( a int ) with ( connector = 'nats', server_url='172.31.46.172:4222', subject='testsubject', connect_mode='plain', consumer.durable_name = 'main_all_3', consumer.ack_policy = 'all', stream='teststream', consumer.max_ack_pending = '1000000') format plain encode json;
```

image

* ack explicit 

```
create table t ( a int ) with ( connector = 'nats', server_url='172.31.46.172:4222', subject='testsubject', connect_mode='plain', consumer.durable_name = 'main_explicit_4', consumer.ack_policy = 'explicit', stream='teststream', consumer.max_ack_pending = '1000000') format plain encode json;
```

image


* ack none

```
create table t ( a int ) with ( connector = 'nats', server_url='172.31.46.172:4222', subject='testsubject', connect_mode='plain', consumer.durable_name = 'main_none_4', consumer.ack_policy = 'none', stream='teststream', consumer.max_ack_pending = '1000000') format plain encode json;
```

image

---

* 4 parallelism + ack all

```
set streaming_parallelism to 4;
create table t ( a int ) with ( connector = 'nats', server_url='172.31.46.172:4222', subject='testsubject', connect_mode='plain', consumer.durable_name = 'branch_all_3', consumer.ack_policy = 'all', stream='teststream', consumer.max_ack_pending = '1000000') format plain encode json;
```

image

* 4 parallelism + ack explicit

```
set streaming_parallelism to 4;
create table t ( a int ) with ( connector = 'nats', server_url='172.31.46.172:4222', subject='testsubject', connect_mode='plain', consumer.durable_name = 'branch_explicit_4', consumer.ack_policy = 'explicit', stream='teststream', consumer.max_ack_pending = '1000000') format plain encode json;
```

image

* 4 parallelism + ack none

```
set streaming_parallelism to 4;
create table t ( a int ) with ( connector = 'nats', server_url='172.31.46.172:4222', subject='testsubject', connect_mode='plain', consumer.durable_name = 'branch_none_3', consumer.ack_policy = 'none', stream='teststream', consumer.max_ack_pending = '1000000') format plain encode json;
```

image

File: src/meta/src/stream/source_manager.rs [hunk 1]
```diff
@@ -520,13 +535,17 @@ impl<T: SplitMetaData + Clone> Ord for ActorSplitsAssignment<T> {
 #[derive(Debug)]
 struct SplitDiffOptions {
     enable_scale_in: bool,
+
+    /// For most connectors, this should be false. When enabled, RisingWave will not track any progress.
+    enable_adaptive: bool,
```
[the remark this answers is not in the document]
Comment (fuyufjh) [reply src/meta/src/stream/source_manager.rs:540]: How about renaming `enable_scale_in` to `enable_drop_splits`?

File: src/meta/src/stream/source_manager.rs [hunk 2]
```diff
@@ -326,7 +333,7 @@ impl SourceManagerCore {
             };
             let backfill_fragment_ids = self.backfill_fragments.get(source_id);
 
-            let Some(discovered_splits) = handle.discovered_splits().await else {
+            let Some(mut discovered_splits) = handle.discovered_splits().await else {
```
Comment (fuyufjh) [inline src/meta/src/stream/source_manager.rs:336]:
```suggestion
            let Some(discovered_splits) = if !handle.enable_adaptive_splits {
                let splits = handle.discovered_splits().await;
                debug_assert!(discovered_splits.len() == 1);
                let split = discovered_splits.values().next().unwrap();
                fill_adaptive_split(split, actors)
            }
```

File: src/connector/src/source/util.rs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+use crate::error::{ConnectorError, ConnectorResult};
+use crate::source::nats::split::NatsSplit;
+use crate::source::SplitImpl;
+
+pub fn fill_adaptive_split(
+    discovered_splits: BTreeMap<Arc<str>, SplitImpl>,
+    actor_in_use: &HashSet<u32>,
+) -> ConnectorResult<BTreeMap<Arc<str>, SplitImpl>> {
+    debug_assert!(discovered_splits.len() == 1);
+    let split = discovered_splits.values().next().unwrap();
```
Comment (fuyufjh) [inline src/connector/src/source/util.rs:27]: In that case, please just pass one split in.

```suggestion
pub fn fill_adaptive_split(
    discovered_split: SplitImpl,
    actor_in_use: &HashSet<u32>,
) -> ConnectorResult<BTreeMap<Arc<str>, SplitImpl>> {
```

**Review by fuyufjh — APPROVED** [review]

**Pull request merged**